bastionzero/zli
# Send logs to S3 Bucket

## Description of the change

This pull request is for the `zli send-logs` feature. This feature allows users to package their zli, daemon, and agent logs (for a specific bzero target) and send them to an S3 bucket managed by the BastionZero dev team. 

The feature will send logs from the current day (the day the command is run), and the previous day (filtered by the datestamp in each log). This will allow us to have anywhere between the past 24-48 hours of logs. 

**Helpful note:** Last two days means that if I run the command today on Oct 4th, then all the logs starting from 12:00am Oct 3rd until this moment will be retrieved, which is why I listed it as the past 24-48 hours. (Query logs for Oct 3 and Oct 4)

The filesystem in the S3 bucket is as follows:
`Month/Customer_Email_Domain/Customer_Email/Request_UUID/[your posted zipped log files]`

This feature is supported by the following commands:
`zli send-logs` which will zip the zli and daemon logs stored on the user's machine and post them to S3. 
`zli send-logs --target [some_agent_target]` which will zip the bzero/kube logs stored on the agent `some_agent_target` and post them to S3. 
`zli send-logs --all [some_agent_target]` which will do a conjunction of both previous commands. 

## Testing

**Test 1**
1. Run `zlil configure` and make sure that you have logs existing in your daemon and zli dev files to post to S3.
2. Run `zli send-logs`.
4. Expect to see the unique UUID for this request printed to the console and also to see a slack notification in `#customer-logs-alarms`. The slack notification will also contain this UUID for reference along with the full object path.
5. Go to the `bastionzero-dev-customer-logs` S3 bucket in AWS and find your files using the filesystem listed above for reference.
6. Download the zip files, unzip them and expect to see all **zli and daemon** logs, timestamped with the last two days. 

 **Test 2**
1. Run `zli send-logs --target [some_agent_target]`.
3. Follow steps 3 and 4 listed for Test 1
4. Download the zip files, unzip them and expect to see all **agent** logs for the target you specified. You should see the current `bzero-agent.log` file and also up to 2 rotated log files for bzero agents. The explanation for how rotated log files are handled can be found in [the bzero pull request](https://github.com/bastionzero/bzero/pull/153). You should be able to see a log file with logs from the past 48 hours for kube agents. 

 **Test 3**
1. Run `zlil configure` and make sure that you have logs existing in your daemon and zli dev files to post to S3.
2. Run `zli send-logs --all [some_agent_target]`.
3. Follow steps 3 and 4 listed for Test 1
7. Download the zip files, unzip them and expect to see the **zli, daemon, and agent** logs, timestamped with the last two days. This command is essentially a combination of test 1 and 2. You can expect to see the zip file for agent logs and zip file for zli and daemon logs grouped under the same folder labeled with the UUID for this request. 


**backend branch:** send-logs-to-S3
**bzero branch:** develop
**charts branch:**

#### Ready to run system tests?


## Relevant release note information

Release Notes: Allows users to send logs to BastionZero

## Related JIRA tickets

Relates to JIRA: CWC-2024

## Relevant HotFix if applicable

Relevant HotFix PR Number:

## Have you considered the security impacts?

Does this PR have any security impact?


If yes, please explain:
